### ISSUE TITLE
Pyright CLI and VSCode problem tab gives different results - total_ordering

### ISSUE THREAD

**Issue by matheuskknd** [microsoft/pylance-release#2989]
# Environment data

* Language Server version: Pylance version 2022.6.31, VSCode 1.68.1 (user setup)
* OS: Windows 10 Pro, Version 21H2, Build 19044.1766
* Python version: Python 3.10.5
* Pyright CLI version: pyright 1.1.256

This issue is very similar to the one presented by Pierre Snell (@Ierezell) in the issue #2255. I just opened this one in a separete thread due to the strange interaction of the issue with ```functools.total_ordering```.

# Expected behaviour

Having the same problems in the "problem" tab and when running ````pyright```` by hand.

# Actual behaviour

The problems tab on Visual Studio Code (VSCode) shows no information, warning nor error message. But if I try to run the following command on the project directory: ````pyright --project pyrightconfig.json```` (by hand) indeed it gives "new" errors (not shown in the problems tab).

Note, however, that if I remove the completelly the usage of ```functools.total_ordering``` from the script, the behaviour is the expected one.

# Project directory

```bash
my.user@PC MINGW64 ~/Desktop/pyright_issue_2989_sample (master)
$ ls -lha
total 26K
drwxr-xr-x 1 matheus.delgado 1049089   0 jul  4 16:14 ./
drwxr-xr-x 1 matheus.delgado 1049089   0 jul  4 16:15 ../
drwxr-xr-x 1 matheus.delgado 1049089   0 jul  4 15:53 .git/
-rw-r--r-- 1 matheus.delgado 1049089 226 jul  4 15:32 .gitmodules
-rw-r--r-- 1 matheus.delgado 1049089 715 jul  1 11:46 pyrightconfig.json
-rw-r--r-- 1 matheus.delgado 1049089 163 jul  4 14:33 README.md
-rwxr-xr-x 1 matheus.delgado 1049089 701 jul  4 16:14 snippet.py*
drwxr-xr-x 1 matheus.delgado 1049089   0 jul  4 15:31 typings/
```

# Project submodules

```bash
my.user@PC MINGW64 ~/Desktop/pyright_issue_2989_sample (master)
$ git submodule status
 0aa15ddf9f6ec4c38b822306d3e05b45e22b6680 typings/python-type-stubs (heads/main)
 7b54854c90aa560d99e353ae2f9fd8b6b9cff40a typings/typeshed (heads/master)
```

# Pyright configuration file

```json
{
  "include": [
    "*.py"
  ],
  "exclude": [
    "**/__pycache__",
    "**/\\.*"
  ],
  "ignore": [
    "typings/*",
    "**/test_*"
  ],
  "defineConstant": {
    "DEBUG": true
  },
  "extraPaths": [],
  "typeshedPath": "./typings/typeshed",
  "stubPath": "./typings/python-type-stubs",
  "typeCheckingMode": "strict",
  "useLibraryCodeForTypes": true,
  "verboseOutput": false,
  "pythonVersion": "3.10",
  "pythonPlatform": "All",
  "reportMissingTypeStubs": "information",
  "reportUnknownVariableType": "information",
  "reportMissingTypeArgument": "information"
}
```

# Python sample code

```python
#!/usr/bin/env python3
# -*- coding: UTF-8 -*-

from typing import TypeAlias, Union
from functools import total_ordering


class Interface:
  pass


InterfaceOrInt: TypeAlias = Union[Interface, int]


@total_ordering
class FooDict(Interface, dict[int, InterfaceOrInt]):

  def foo(self) -> None:
    bar: InterfaceOrInt = self[0]
    if isinstance(bar, (FooDict, FooList)):
      _ = bar.sorted()

  def sorted(self) -> "FooDict":
    return FooDict()

  def __lt__(self, __o: object) -> bool:
    return True


@total_ordering
class FooList(Interface, list[InterfaceOrInt]):

  def sorted(self) -> "FooList":
    return FooList()

```

# Pyright CLI result

```bash
my.user@PC MINGW64 ~/Desktop/Pyright (master)
$ pyright --project pyrightconfig.json
npm WARN config global `--global`, `--local` are deprecated. Use `--location=global` instead.
Loading configuration file at C:\Users\my.user\Desktop\Pyright\pyrightconfig.json
Searching for source files
Found 1 source file
pyright 1.1.256
C:\Users\my.user\Desktop\Pyright\snippet.py
  C:\Users\my.user\Desktop\Pyright\snippet.py:21:21 - error: Cannot access member "sorted" for type "Interface"
    Member "sorted" is unknown (reportGeneralTypeIssues)
  C:\Users\my.user\Desktop\Pyright\snippet.py:21:17 - error: Type of "sorted" is unknown (reportUnknownMemberType)
  C:\Users\my.user\Desktop\Pyright\snippet.py:21:21 - error: Cannot access member "sorted" for type "int"
    Member "sorted" is unknown (reportGeneralTypeIssues)
  C:\Users\my.user\Desktop\Pyright\snippet.py:21:13 - information: Type of "_" is unknown (reportUnknownVariableType)
3 errors, 0 warnings, 1 information
Completed in 1.014sec
```

# VSCode problems tab

![ProofVSCodeProblemsTab](https://user-images.githubusercontent.com/36434509/177208147-16b3b07c-3eee-4adb-8696-67a468ff4f95.png)

**Comment by erictraut** [comment]
I'm not able to repro the problem you're reporting.

When you run the CLI version of pyright, which python interpreter do you have configured in the shell environment? Is it the same one you have selected in the VS Code environment? You might be able to tell by running `python --version` from the command line.

The symptom you're describing leads me to believe that the Python environment you are using the command line has a library (either a stub library or a full package) installed called `functools`, and it is overriding the type information for `total_ordering`.

One thing that might be interesting is to add the line `reveal_type(FooDict)` to the end of your code snippet. It should output something like `Type of "FooDict" is "Type[FooDict]". If you see something other than that, then the `@total_ordering` decorator is returning some type other than the input class, and that would explain the error you're seeing here.

I noticed that you're specifying a custom `typeshedPath`. That means you might be using a version of typeshed that is outdates or inconsistent with the version that ships with pylance and pyright. So that could also explain why I'm not able to repro the problem, although it wouldn't explain why you're seeing different behavior between pylance and the pyright CLI.

**Comment by matheuskknd** [comment]
## First point, enviroments:

### CLI:

```bash
my.user@PC MINGW64 ~
$ python --version
Python 3.10.5
```

### VSCode:

![VsCodeEnv](https://user-images.githubusercontent.com/36434509/177207405-1172f5f5-5914-43b8-b660-d20fcdf1b2f4.JPG)

## Second point, ```reveal_type```:

### CLI:

```bash
my.user@PC MINGW64 ~/Desktop/pyright_issue_2989_sample (main)
$ pyright --project pyrightconfig.json
npm WARN config global `--global`, `--local` are deprecated. Use `--location=global` instead.
Loading configuration file at C:\Users\my.user\Desktop\pyright_issue_2989_sample\pyrightconfig.json
Searching for source files
Found 1 source file
pyright 1.1.256
C:\Users\my.user\Desktop\pyright_issue_2989_sample\snippet.py
  C:\Users\my.user\Desktop\pyright_issue_2989_sample\snippet.py:21:21 - error: Cannot access member "sorted" for type "Interface"
    Member "sorted" is unknown (reportGeneralTypeIssues)
  C:\Users\my.user\Desktop\pyright_issue_2989_sample\snippet.py:21:17 - error: Type of "sorted" is unknown (reportUnknownMemberType)
  C:\Users\my.user\Desktop\pyright_issue_2989_sample\snippet.py:21:21 - error: Cannot access member "sorted" for type "int"
    Member "sorted" is unknown (reportGeneralTypeIssues)
  C:\Users\my.user\Desktop\pyright_issue_2989_sample\snippet.py:21:13 - information: Type of "_" is unknown (reportUnknownVariableType)
  C:\Users\my.user\Desktop\pyright_issue_2989_sample\snippet.py:37:13 - information: Type of "FooDict" is "Type[FooDict]"
3 errors, 0 warnings, 2 informations
Completed in 0.863sec
WARNING: there is a new pyright version available (v1.1.256 -> v1.1.257).
Please install the new version or set PYRIGHT_PYTHON_FORCE_VERSION to `latest`
```

### VSCode:

![ProofPylanceRevealation](https://user-images.githubusercontent.com/36434509/177207354-46b9a32d-0002-4e0f-b3ee-ecebe0639096.JPG)

## Third point:

> I noticed that you're specifying a custom typeshedPath. That means you might be using a version of typeshed that is outdates or inconsistent with the version that ships with pylance and pyright. So that could also explain why I'm not able to repro the problem, although it wouldn't explain why you're seeing different behavior between pylance and the pyright CLI.

I created [this](https://github.com/matheuskknd/pyright_issue_2989_sample) sample git repository. Maybe you are able to reproduce now... The two submodules are the most up to date versions (for my best knowledge) of [Microsoft Python Type Stubs](https://github.com/microsoft/python-type-stubs/tree/main) and [typeshed](https://github.com/python/typeshed/tree/master). But it doesn't matter if I remove both or not, the issue remains the same. I just inserted them for restricting the possibilities of environment errors.

**Comment by erictraut** [comment]
Thanks for preparing the sample git repository. Unfortunately, I am still unable to repro the problem.

I'm running short on theories for why you are seeing this error. My only working hypothesis is that: 1) you are using a different python environment for the CLI and VS Code, and 2) the python environment that you're using for the CLI has an installed package called `functools` (thus overriding the typeshed stub of the same name).

Perhaps you could try to create a fresh virtual environment?
```
python3.10 -m venv venv
source venv/bin/activate
pip install pyright
pyright
```

Here's the output I receive when I do this:
```
Found 1 source file
pyright 1.1.257
/<redacted>/pyright_issue_2989_sample/snippet.py
  /<redacted>/pyright_issue_2989_sample/snippet.py:37:13 - information: Type of "FooDict" is "Type[FooDict]"
0 errors, 0 warnings, 1 information 
Completed in 0.575sec
```

**Comment by matheuskknd** [comment]
Thank you for the quick answer.

Sorry, I did not understand very well where and with what privileges I should run those commands, then I executed them in the sample project directory as my user:

```bash
my.user@PC ~/Desktop/pyright_issue_2989_sample (main)
$ python3.10 -m venv venv

my.user@PC ~/Desktop/pyright_issue_2989_sample (main)
$ source venv/bin/activate
bash: venv/bin/activate: No such file or directory

my.user@PC ~/Desktop/pyright_issue_2989_sample (main)
$ pip install pyright
Requirement already satisfied: pyright in c:\users\my.user\appdata\local\packages\pythonsoftwarefoundation.python.3.10_qbz5n2kfra8p0\localcache\local-packages\python310\site-packages (1.1.256)
Requirement already satisfied: nodeenv>=1.6.0 in c:\users\my.user\appdata\local\packages\pythonsoftwarefoundation.python.3.10_qbz5n2kfra8p0\localcache\local-packages\python310\site-packages (from pyright) (1.7.0)
Requirement already satisfied: setuptools in c:\program files\windowsapps\pythonsoftwarefoundation.python.3.10_3.10.1520.0_x64__qbz5n2kfra8p0\lib\site-packages (from nodeenv>=1.6.0->pyright) (58.1.0)
WARNING: There was an error checking the latest version of pip.

my.user@PC ~/Desktop/pyright_issue_2989_sample (main)
$ pyright
npm WARN config global `--global`, `--local` are deprecated. Use `--location=global` instead.
Loading configuration file at C:\Users\my.user\Desktop\pyright_issue_2989_sample\pyrightconfig.json
Searching for source files
Found 1 source file
pyright 1.1.256
C:\Users\my.user\Desktop\pyright_issue_2989_sample\snippet.py
  C:\Users\my.user\Desktop\pyright_issue_2989_sample\snippet.py:21:21 - error: Cannot access member "sorted" for type "Interface"
    Member "sorted" is unknown (reportGeneralTypeIssues)
  C:\Users\my.user\Desktop\pyright_issue_2989_sample\snippet.py:21:17 - error: Type of "sorted" is unknown (reportUnknownMemberType)
  C:\Users\my.user\Desktop\pyright_issue_2989_sample\snippet.py:21:21 - error: Cannot access member "sorted" for type "int"
    Member "sorted" is unknown (reportGeneralTypeIssues)
  C:\Users\my.user\Desktop\pyright_issue_2989_sample\snippet.py:21:13 - information: Type of "_" is unknown (reportUnknownVariableType)
  C:\Users\my.user\Desktop\pyright_issue_2989_sample\snippet.py:37:13 - information: Type of "FooDict" is "Type[FooDict]"
3 errors, 0 warnings, 2 informations
Completed in 0.812sec
WARNING: there is a new pyright version available (v1.1.256 -> v1.1.257).
Please install the new version or set PYRIGHT_PYTHON_FORCE_VERSION to `latest`
```

Note also that I'm using Windows 10 with GitBash (MINGW64). But, again, even if I run these commands in the Windows CMD the erros are amost the same (except for ```source``` not being a known command).

**Comment by erictraut** [comment]
Yes, I should have specified that you should run these commands within your project directory and with your user permissions. So you made the correct assumptions.

It looks like you didn't activate the venv in step 2. On Windows, you will need to run `venv\Scripts\activate.bat` to activate the venv. If it is correctly activated, you will not see an error message like "No such file or directory" after this step.

**Comment by matheuskknd** [comment]
Now it seems to work:

```cmd
C:\Users\my.user\Desktop\pyright_issue_2989_sample>python3.10 -m venv venv

C:\Users\my.user\Desktop\pyright_issue_2989_sample>venv\Scripts\activate.bat

(venv) C:\Users\my.user\Desktop\pyright_issue_2989_sample>pip install pyright
Collecting pyright
  Downloading pyright-1.1.257-py3-none-any.whl (16 kB)
Collecting nodeenv>=1.6.0
  Using cached nodeenv-1.7.0-py2.py3-none-any.whl (21 kB)
Requirement already satisfied: setuptools in c:\users\my.user\desktop\pyright_issue_2989_sample\venv\lib\site-packages (from nodeenv>=1.6.0->pyright) (58.1.0)
Installing collected packages: nodeenv, pyright
Successfully installed nodeenv-1.7.0 pyright-1.1.257
WARNING: You are using pip version 22.0.4; however, version 22.1.2 is available.
You should consider upgrading via the 'C:\Users\my.user\Desktop\pyright_issue_2989_sample\venv\Scripts\python.exe -m pip install --upgrade pip' command.

(venv) C:\Users\my.user\Desktop\pyright_issue_2989_sample>pyright
npm WARN config global `--global`, `--local` are deprecated. Use `--location=global` instead.
Loading configuration file at C:\Users\my.user\Desktop\pyright_issue_2989_sample\pyrightconfig.json
Searching for source files
Found 1 source file
pyright 1.1.257
C:\Users\my.user\Desktop\pyright_issue_2989_sample\snippet.py
  C:\Users\my.user\Desktop\pyright_issue_2989_sample\snippet.py:37:13 - information: Type of "FooDict" is "Type[FooDict]"
0 errors, 0 warnings, 1 information
Completed in 1.036sec
```

Thank you very much, despite I cannot have the same output with GitBash (MINGW64). Do you have some matterial to help understanding what is Python environments and how to setup it properly for working with pyright?

**Comment by erictraut** [comment]
How to set up and manage Python environments is beyond the scope of the pylance forum. There are many good resources on the internet that can explain the basic concepts. Just search for "Python virtual environment". Here's a good [starting point](https://docs.python.org/3/tutorial/venv.html).

**Comment by matheuskknd** [comment]
Thank you!

**Comment by matheuskknd** [comment]
Hello again.

I am reopening the issue just because it may be interesting for someone to know how I solved the initial issue in the real code. And because I think I found a similar issue.

After studying a little about [Python Environments](https://docs.python.org/3/tutorial/venv.html) as suggested above, I was able to create an environment in my real project. At the same time, I also learned how to "pin" the dependencies and it's versions onto the project using ```requirements.txt``` files. These are used for the initial setup of a recently created environment. Examples can be found in the [sample repository](https://github.com/matheuskknd/pyright_issue_2989_sample).

When I finally finished studying and implementing the changes, I runned ```pyright``` (by hand) and accidentally hit another similar issue. This time, Pyright popped out only one point of error, which I isolated and uploaded to the [sample repository](https://github.com/matheuskknd/pyright_issue_2989_sample) in a file named ```snippet02.py```. The ```requirements.txt``` file is also there.

Then, after creating the environment as follows:

```bash
my.user@PC MINGW64 ~/Desktop/pyright_issue_2989_sample (main)
$ python3 -m venv .venv

my.user@PC MINGW64 ~/Desktop/pyright_issue_2989_sample (main)
$ source .venv/Scripts/activate
(.venv)
my.user@PC MINGW64 ~/Desktop/pyright_issue_2989_sample (main)
$ python -m pip install -r requirements.txt
Collecting nodeenv==1.7.0
  Using cached nodeenv-1.7.0-py2.py3-none-any.whl (21 kB)
Collecting pyright==1.1.257
  Using cached pyright-1.1.257-py3-none-any.whl (16 kB)
Requirement already satisfied: setuptools in c:\users\my.user\desktop\pyright_issue_2989_sample\.venv\lib\site-packages (from nodeenv==1.7.0->-r requirements.txt (line 7)) (58.1.0)
Installing collected packages: nodeenv, pyright
Successfully installed nodeenv-1.7.0 pyright-1.1.257
(.venv)
```

I got...

## Pyright CLI result

```bash
my.user@PC MINGW64 ~/Desktop/pyright_issue_2989_sample (main)
$ pyright --project pyrightconfig.json snippet02.py
npm WARN config global `--global`, `--local` are deprecated. Use `--location=global` instead.
Loading configuration file at C:\Users\my.user\Desktop\pyright_issue_2989_sample\pyrightconfig.json
Searching for source files
Found 1 source file
pyright 1.1.257
C:\Users\my.user\Desktop\pyright_issue_2989_sample\snippet02.py
  C:\Users\my.user\Desktop\pyright_issue_2989_sample\snippet02.py:38:26 - error: Cannot access member "bar" for type "<subclass of int and FooBase>"
    Member "bar" is unknown (reportGeneralTypeIssues)
  C:\Users\my.user\Desktop\pyright_issue_2989_sample\snippet02.py:38:20 - error: Type of "bar" is unknown (reportUnknownMemberType)
  C:\Users\my.user\Desktop\pyright_issue_2989_sample\snippet02.py:38:20 - error: Type of "bar" is partially unknown
    Type of "bar" is "(() -> str) | Unknown" (reportUnknownMemberType)
  C:\Users\my.user\Desktop\pyright_issue_2989_sample\snippet02.py:38:20 - information: Return type, "str | Unknown", is partially unknown (reportUnknownVariableType)
  C:\Users\my.user\Desktop\pyright_issue_2989_sample\snippet02.py:53:13 - information: Type of "Foo" is "Type[Foo]"
3 errors, 0 warnings, 2 informations
Completed in 0.659sec
(.venv)
```

## VSCode problems tab

![VSCodeProblemsTabProof](https://user-images.githubusercontent.com/36434509/177418180-0ae63973-ef3d-4a6c-b1a7-3391a60f7a89.png)

## Summary

The ```snippet02.py``` file contains the member function ```Foo.notWorkingFoo``` which generates the error. Once the real code type checking only broke at this point, I just changed the implementation to be like the one in ```Foo.workingFoo_2``` that has the same semantics.

The point here is that the behaviour is not the same again between VSCode Pylance and CLI Pyright. And, again, if ```total_ordering``` is completelly removed, the behaviour is the same.

*Notes*: no error is shown in ```Foo.workingFoo_1``` member function while the class is decorated with ```total_ordering```. But, if removed, an error is raised by both VSCode Pylance and Pyright.

**Comment by heejaechang** [comment]
This issue has been fixed in version 2022.7.21, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#2022721-6-July-2022)
